src/Compilers/Core/Portable/PEWriter/PeWriter.cs

```diff
@@ -35,13 +33,69 @@ public PeWritingException(Exception inner)
 
     internal static class PeWriter
     {
+        internal struct EmitBuilders
+        {
+            internal BlobBuilder IlBlobBuilder;
+            internal PooledBlobBuilder MappedFieldDataBlobBuilder;
+            internal PooledBlobBuilder ManagedResourceBlobBuilder;
+            internal PooledBlobBuilder? PortableExecutableBlobBuilder;
+            internal PooledBlobBuilder? PortablePdbBlobBuilder;
+
+            public EmitBuilders()
+            {
+                IlBlobBuilder = new BlobBuilder(32 * 1024);
+                MappedFieldDataBlobBuilder = PooledBlobBuilder.GetInstance();
+                ManagedResourceBlobBuilder = PooledBlobBuilder.GetInstance();
+                PortableExecutableBlobBuilder = null;
+                PortablePdbBlobBuilder = null;
+            }
+
+            internal void Free()
+            {
+                if (PortableExecutableBlobBuilder is null)
+                {
+                    MappedFieldDataBlobBuilder.Free();
+                    ManagedResourceBlobBuilder.Free();
+                }
+                else
+                {
+                    var mappedCount = MappedFieldDataBlobBuilder.Count;
+                    var resourceCount = ManagedResourceBlobBuilder.Count;
+
+                    PortableExecutableBlobBuilder.Free();
+
+                    // Once PortableExecutableBuilder is created it becomes the owner of the 
+                    // MappedFieldDataBuilder and ManagedResourceBuilder instances. Freeing 
+                    // it is sufficient to free both of them.
+                    //
+                    // However there is a bug in LinkSuffix / LinkPrefix which causes the
+                    // ownership to not happen when these are length 0. Need to handle that
+                    // specially until this is fixed.
+                    // https://github.com/dotnet/runtime/issues/99266
+                    if (mappedCount == 0)
+                    {
+                        MappedFieldDataBlobBuilder.AssertAlive();
```

Comment:
In a previous iteration of this change I had a similar method called `AssertFreed` and @333fred rightfully pointed out it's a race condition. As written now this call is not a race cause the `BlobBuilder` will be alive. Once the bug mentioned here is fixed this assert also becomes a race condition. However the chance of the race is unlikely enough that this will still serve as an sufficient trip wire to ensure that we don't forget to update this code.